Update K2 to pass `make check`

Risca opened the pull request:
Summary of changes:
* A stub `k2-multirunner` addon has been added. It doesn't provide any extensions; only the definitions of `MULTI_RUNNER_ENABLED`, `MULTI_RUNNER_ENDPOINT`, and `TEST_SUBRUN`.
* System tests that rely on the "multirunner" extension have been skipped. The ExtensionMananger component is used to check for the existence of the "multirunner" extension so if `k2-multirunner` gets re-introduced, these tests should automatically get re-enabled again. The addons that depend on `k2-multirunner` are:
  * `k2-consolerunner`
  * `k2-gtestbinaryrunner`
* The order of attributes in an XML tag changes between python and/or OS versions, so a third party library was introduced to help the unit tests pass for the test report writers.
* Various dependencies have been updated to newer versions due to old versions no longer being available at pypi.
* Other minor fixes

It is recommended that each commit be reviewed individually.

These changes have been tested on both Ubuntu server 18.04 (python 3.6) and Arch Linux (python 3.9)

Risca commented:
@per-bohlin, if you manage to find the Gude code, I'm happy to take over as maintainer of the `k2-power` addon since I've written fair bit of that code originally :wink:

andni233 commented:
LGTM

per-bohlin commented:
@Risca , I think I will be able to dig up more code.
The code here is just a first publication. I have access to more open source code from upstream that wasn't published in the first iteration. I've had the plan for a very long time to work on it and publish it properly but never got around to it. I will go through it in more detail now that I know someone wants to use it. At least to publish it as is.

>     * A stub `k2-multirunner` addon has been added. It doesn't provide any extensions; only the definitions of `MULTI_RUNNER_ENABLED`, `MULTI_RUNNER_ENDPOINT`, and `TEST_SUBRUN`.
I have access to the multirunner. I will publish it before end of year.

>     * Same thing with the power meter and power switch systests: they are lacking the Gude power meter and power switch, and have been skipped.
It is unfortunate if the general power-meter and power-switch is dependent on the specific implementation of the Gude power control. That should be changed so they use the dummy instead - if they need a specific implementation at all.
That aside, I have access to the Gude Power Control source and will publish it before end of year.

per-bohlin commented:
Given that this clashes with the intended publication of the multi-runner and gude-power-controller, it would be great if you can hold off and rebase this MR once they have been published. I'll write a message in this review when it is done.

Risca commented:
> It is unfortunate if the general power-meter and power-switch is dependent on the specific implementation of the Gude power control. That should be changed so they use the dummy instead - if they need a specific implementation at all.

Agreed. The system tests are more or less system tests of the Gude code, and not the power meter/switch interfaces. I'll see what I can do about them, but don't get your hopes up - it's not really at the top of my list of priorities :wink: 

> Given that this clashes with the intended publication of the multi-runner and gude-power-controller, it would be great if you can hold off and rebase this MR once they have been published. I'll write a message in this review when it is done.

I'll hold :smile_cat:

Risca commented:
Oh, I almost forgot to mention: I haven't got the infrastructure setup to run the docker systests, so they are listed as "SKIPPED" for me.

per-bohlin commented:
I just pushed K2 Gude Power Control addon ( #16 ).
Multi-runner coming soon.

Risca commented:
> > It is unfortunate if the general power-meter and power-switch is dependent on the specific implementation of the Gude power control. That should be changed so they use the dummy instead - if they need a specific implementation at all.
> 
> Agreed. The system tests are more or less system tests of the Gude code, and not the power meter/switch interfaces. I'll see what I can do about them, but don't get your hopes up - it's not really at the top of my list of priorities 

Now that the Gude power control addon is pushed, I reverted my changes in that systest. I created a separate pull request (#18) that improves the situation slightly.

per-bohlin commented:
@Risca , FYI: multirunner has been merged (#20).

per-bohlin reviewed (approved):
Looks good. Thanks! Merge at will.

Risca commented:
Rebased on latest master and updated! `make check` on Arch Linux and Ubuntu server 18.04 passes. There is currently a race condition in one of the telnet systests that should be solved by #21 and the the latest refactoring I've done.

Risca commented:
> Rebased on latest master and updated! `make check` on Arch Linux and Ubuntu server 18.04 passes. There is currently a race condition in one of the telnet systests that should be solved by #21 and the the latest refactoring I've done.

I'll push the refactored systest as a separate pull request (#22) instead. I also had some problems when I had a locally installed flake8 in `~/.local/bin`. Looks like my local environment got priority over the venv paths. Solved by `pip uninstall flake8` :shrug: (and out of scope for this PR)
